File: app/iosApp/iosApp/NativeModules/NativeWebSocketModule.swift
import Foundation
import UIKit

private final class NativeWebSocketConnection {
  let id: String
  let task: URLSessionWebSocketTask
  var terminal = false
  var closing = false
  var requestedCloseCode = 1000
  var requestedCloseReason = ""

  init(id: String, task: URLSessionWebSocketTask) {
    self.id = id
    self.task = task
  }
}

/// Page-owned production WebSocket transport independent of Lynx DevTool.
@objcMembers
final class NativeWebSocketController: NSObject, URLSessionWebSocketDelegate {
  static let eventName = "nativeWebSocket"

  private weak var lynxView: LynxView?
  private let lock = NSLock()
  private var connectionsByID: [String: NativeWebSocketConnection] = [:]
  private var connectionsByTaskID: [Int: NativeWebSocketConnection] = [:]
  private var destroyed = false
  private var sessionWasUsed = false

  private lazy var delegateQueue: OperationQueue = {
    let queue = OperationQueue()
    queue.name = "com.lynxapp.websocket"
    queue.maxConcurrentOperationCount = 1
    return queue
  }()

  private lazy var session: URLSession = URLSession(
    configuration: .default,
    delegate: self,
    delegateQueue: delegateQueue
  )

  func attach(lynxView: LynxView) {
    self.lynxView = lynxView
  }

  func connect(
    id: String,
    url: String,
    protocols: [String],
    headers: [String: String]
  ) -> String {
    guard id.range(
      of: "^[A-Za-z0-9._-]{1,128}$",
      options: .regularExpression
    ) != nil else {
      return "Invalid WebSocket connection ID"
    }
    guard isAllowedURL(url), let targetURL = URL(string: url) else {
      return "WebSocket URL must use wss:// (ws:// is Debug-only)"
    }

    lock.lock()
    let canConnect = !destroyed && connectionsByID[id] == nil
    lock.unlock()
    guard canConnect else {
      return destroyed
        ? "WebSocket host has been destroyed"
        : "WebSocket connection ID already exists"
    }

    var request = URLRequest(url: targetURL)
    for (name, value) in headers {
      request.addValue(value, forHTTPHeaderField: name)
    }
    if !protocols.isEmpty {
      request.setValue(
        protocols.joined(separator: ", "),
        forHTTPHeaderField: "Sec-WebSocket-Protocol"
      )
    }
    sessionWasUsed = true
    let task = session.webSocketTask(with: request)
    let connection = NativeWebSocketConnection(id: id, task: task)

    lock.lock()
    guard !destroyed, connectionsByID[id] == nil else {
      lock.unlock()
      task.cancel()
      return destroyed
        ? "WebSocket host has been destroyed"
        : "WebSocket connection ID already exists"
    }
    connectionsByID[id] = connection
    connectionsByTaskID[task.taskIdentifier] = connection
    lock.unlock()

    task.resume()
    receiveNext(connection)
    return ""
  }

  func sendText(
    id: String,
    data: String,
    completion: @escaping (String) -> Void
  ) {
    send(id: id, message: .string(data), completion: completion)
  }

  func sendBase64(
    id: String,
    data: String,
    completion: @escaping (String) -> Void
  ) {
    guard let decoded = Data(base64Encoded: data) else {
      completion("Invalid Base64 WebSocket payload")
      return
    }
    send(id: id, message: .data(decoded), completion: completion)
  }

  func close(id: String, code: Int, reason: String) -> String {
    guard let connection = connection(forID: id) else {
      return "Unknown WebSocket connection ID"
    }
    guard let closeCode = URLSessionWebSocketTask.CloseCode(rawValue: code) else {
      return "Invalid WebSocket close code"
    }
    connection.closing = true
    connection.requestedCloseCode = code
    connection.requestedCloseReason = reason
    connection.task.cancel(
      with: closeCode,
      reason: reason.data(using: .utf8)
    )
    return ""
  }

  func destroy() {
    lock.lock()
    guard !destroyed else {
      lock.unlock()
      return
    }
    destroyed = true
    let connections = Array(connectionsByID.values)
    connectionsByID.removeAll()
    connectionsByTaskID.removeAll()
    lynxView = nil
    lock.unlock()

    for connection in connections {
      connection.task.cancel(with: .goingAway, reason: nil)
    }
    if sessionWasUsed {
      session.invalidateAndCancel()
    }
  }

  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didOpenWithProtocol protocol: String?
  ) {
    guard let connection = connection(forTaskID: webSocketTask.taskIdentifier) else {
      return
    }
    emit(id: connection.id, payload: [
      "type": "open",
      "protocol": `protocol` ?? "",
    ])
  }

  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
    reason: Data?
  ) {
    guard let connection = takeConnection(forTaskID: webSocketTask.taskIdentifier) else {
      return
    }
    emitClose(
      id: connection.id,
      code: closeCode.rawValue,
      reason: reason.flatMap { String(data: $0, encoding: .utf8) } ?? "",
      wasClean: true
    )
  }

  func urlSession(
    _ session: URLSession,
    task: URLSessionTask,
    didCompleteWithError error: Error?
  ) {
    guard let error,
          let connection = connection(forTaskID: task.taskIdentifier) else {
      return
    }
    if connection.closing,
       (error as NSError).domain == NSURLErrorDomain,
       (error as NSError).code == NSURLErrorCancelled {
      guard takeConnection(forTaskID: task.taskIdentifier) != nil else { return }
      emitClose(
        id: connection.id,
        code: connection.requestedCloseCode,
        reason: connection.requestedCloseReason,
        wasClean: true
      )
      return
    }
    guard takeConnection(forTaskID: task.taskIdentifier) != nil else { return }
    emitFailure(id: connection.id, message: error.localizedDescription)
  }

  private func send(
    id: String,
    message: URLSessionWebSocketTask.Message,
    completion: @escaping (String) -> Void
  ) {
    guard let connection = connection(forID: id) else {
      completion("Unknown WebSocket connection ID")
      return
    }
    connection.task.send(message) { error in
      completion(error?.localizedDescription ?? "")
    }
  }

  private func receiveNext(_ connection: NativeWebSocketConnection) {
    connection.task.receive { [weak self, weak connection] result in
      guard let self, let connection,
            self.connection(forID: connection.id) === connection else {
        return
      }
      switch result {
      case let .success(message):
        switch message {
        case let .string(value):
          self.emit(id: connection.id, payload: [
            "type": "message",
            "dataType": "text",
            "data": value,
          ])
        case let .data(value):
          self.emit(id: connection.id, payload: [
            "type": "message",
            "dataType": "base64",
            "data": value.base64EncodedString(),
          ])
        @unknown default:
          self.emitFailure(
            id: connection.id,
            message: "Unsupported WebSocket message type"
          )
          return
        }
        self.receiveNext(connection)
      case let .failure(error):
        if connection.closing { return }
        guard self.takeConnection(forID: connection.id) != nil else { return }
        self.emitFailure(id: connection.id, message: error.localizedDescription)
      }
    }
  }

  private func connection(forID id: String) -> NativeWebSocketConnection? {
    lock.lock()
    defer { lock.unlock() }
    return destroyed ? nil : connectionsByID[id]
  }

  private func connection(forTaskID id: Int) -> NativeWebSocketConnection? {
    lock.lock()
    defer { lock.unlock() }
    return destroyed ? nil : connectionsByTaskID[id]
  }

  private func takeConnection(forID id: String) -> NativeWebSocketConnection? {
    lock.lock()
    defer { lock.unlock() }
    guard !destroyed, let connection = connectionsByID.removeValue(forKey: id) else {
      return nil
    }
    connectionsByTaskID.removeValue(forKey: connection.task.taskIdentifier)
    connection.terminal = true
    return connection
  }

  private func takeConnection(forTaskID id: Int) -> NativeWebSocketConnection? {
    lock.lock()
    defer { lock.unlock() }
    guard !destroyed, let connection = connectionsByTaskID.removeValue(forKey: id) else {
      return nil
    }
    connectionsByID.removeValue(forKey: connection.id)
    connection.terminal = true
    return connection
  }

  private func emitFailure(id: String, message: String) {
    emit(id: id, payload: [
      "type": "error",
      "message": message,
    ])
    emitClose(id: id, code: 1006, reason: message, wasClean: false)
  }

  private func emitClose(
    id: String,
    code: Int,
    reason: String,
    wasClean: Bool
  ) {
    emit(id: id, payload: [
      "type": "close",
      "code": code,
      "reason": reason,
      "wasClean": wasClean,
    ])
  }

  private func emit(id: String, payload: [String: Any]) {
    var event = payload
    event["id"] = id
    DispatchQueue.main.async { [weak self] in
      guard let self else { return }
      self.lock.lock()
      let canEmit = !self.destroyed
      self.lock.unlock()
      if canEmit {
        self.lynxView?.sendGlobalEvent(Self.eventName, withParams: [event])
      }
    }
  }

  private func isAllowedURL(_ value: String) -> Bool {
    guard let components = URLComponents(string: value),
          components.host != nil,
          let scheme = components.scheme?.lowercased() else {
      return false
    }
    if scheme == "wss" { return true }
    #if DEBUG
    return scheme == "ws"
    #else
    return false
    #endif
  }
}

/// Lynx NativeModule facade for the app-owned iOS WebSocket controller.
@objcMembers
final class NativeWebSocketModule: NSObject, LynxModule {
  static let name = "NativeWebSocketModule"
  static let methodLookup = [
    "connect": "connect:callback:",
    "sendText": "sendText:data:callback:",
    "sendBase64": "sendBase64:data:callback:",
    "close": "close:code:reason:callback:",
  ]

  private weak var controller: NativeWebSocketController?

  @objc(initWithParam:)
  init(param: Any) {
    controller = param as? NativeWebSocketController
    super.init()
  }

  override init() {
    super.init()
  }

  func connect(_ options: NSDictionary, callback: LynxCallbackBlock) {
    guard let controller else {
      callback("Native WebSocket has no UIViewController host")
      return
    }
    let id = options["id"] as? String ?? ""
    let url = options["url"] as? String ?? ""
    let protocols = options["protocols"] as? [String] ?? []
    let headers = options["headers"] as? [String: String] ?? [:]
    callback(controller.connect(
      id: id,
      url: url,
      protocols: protocols,
      headers: headers
    ))
  }

  func sendText(
    _ id: String,
    data: String,
    callback: @escaping LynxCallbackBlock
  ) {
    guard let controller else {
      callback("Native WebSocket has no UIViewController host")
      return
    }
    controller.sendText(id: id, data: data, completion: callback)
  }

  func sendBase64(
    _ id: String,
    data: String,
    callback: @escaping LynxCallbackBlock
  ) {
    guard let controller else {
      callback("Native WebSocket has no UIViewController host")
      return
    }
    controller.sendBase64(id: id, data: data, completion: callback)
  }

  func close(
    _ id: String,
    code: Int,
    reason: String,
    callback: LynxCallbackBlock
  ) {
    callback(
      controller?.close(id: id, code: code, reason: reason)
        ?? "Native WebSocket has no UIViewController host"
    )
  }
}

File: app/androidApp/app/src/main/java/com/lynxapp/nativemodule/NativeWebSocketModule.kt
package com.lynxapp.nativemodule

import android.app.Activity
import android.content.Context
import android.util.Base64
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.react.bridge.JavaOnlyMap
import com.lynx.react.bridge.ReadableMap
import com.lynx.tasm.LynxView
import com.lynxapp.AppHttpClient
import com.lynxapp.BuildConfig
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import okio.ByteString.Companion.toByteString

/** Activity-owned, production WebSocket transport independent of Lynx DevTool. */
class NativeWebSocketController(private val activity: Activity) {
    private class Connection {
        val terminal = AtomicBoolean(false)

        @Volatile
        var socket: WebSocket? = null
    }

    private val connections = ConcurrentHashMap<String, Connection>()

    @Volatile
    private var lynxView: LynxView? = null

    @Volatile
    private var destroyed = false

    fun attach(lynxView: LynxView) {
        this.lynxView = lynxView
    }

    fun connect(
        id: String,
        url: String,
        protocols: List<String>,
        headers: Map<String, String>,
    ): String {
        if (destroyed) return "WebSocket host has been destroyed"
        if (!CONNECTION_ID.matches(id)) return "Invalid WebSocket connection ID"
        if (!isAllowedUrl(url)) return "WebSocket URL must use wss:// (ws:// is Debug-only)"
        if (connections.containsKey(id)) return "WebSocket connection ID already exists"

        val request = runCatching {
            Request.Builder()
                .url(url)
                .apply {
                    headers.forEach { (name, value) -> addHeader(name, value) }
                    if (protocols.isNotEmpty()) {
                        header("Sec-WebSocket-Protocol", protocols.joinToString(", "))
                    }
                }
                .build()
        }.getOrElse { error ->
            return error.message ?: "Invalid WebSocket request"
        }

        val connection = Connection()
        connections[id] = connection
        connection.socket = runCatching {
            AppHttpClient.instance.newWebSocket(
                request,
                object : WebSocketListener() {
                    override fun onOpen(webSocket: WebSocket, response: Response) {
                        emit(
                            id,
                            JavaOnlyMap().apply {
                                putString("type", "open")
                                putString(
                                    "protocol",
                                    response.header("Sec-WebSocket-Protocol").orEmpty(),
                                )
                            },
                        )
                    }

                    override fun onMessage(webSocket: WebSocket, text: String) {
                        emit(
                            id,
                            JavaOnlyMap().apply {
                                putString("type", "message")
                                putString("dataType", "text")
                                putString("data", text)
                            },
                        )
                    }

                    override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                        emit(
                            id,
                            JavaOnlyMap().apply {
                                putString("type", "message")
                                putString("dataType", "base64")
                                putString("data", bytes.base64())
                            },
                        )
                    }

                    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                        finish(id, connection, code, reason, true)
                    }

                    override fun onFailure(
                        webSocket: WebSocket,
                        t: Throwable,
                        response: Response?,
                    ) {
                        if (!connection.terminal.compareAndSet(false, true)) return
                        connections.remove(id, connection)
                        val message = t.message ?: "WebSocket transport failed"
                        emit(
                            id,
                            JavaOnlyMap().apply {
                                putString("type", "error")
                                putString("message", message)
                            },
                        )
                        emitClose(id, 1006, message, false)
                    }
                },
            )
        }.getOrElse { error ->
            connections.remove(id, connection)
            return error.message ?: "Unable to create WebSocket"
        }
        return ""
    }

    fun sendText(id: String, data: String): String = send(id) { socket ->
        socket.send(data)
    }

    fun sendBase64(id: String, data: String): String {
        val bytes = runCatching { Base64.decode(data, Base64.DEFAULT) }
            .getOrElse { return "Invalid Base64 WebSocket payload" }
        return send(id) { socket -> socket.send(bytes.toByteString()) }
    }

    fun close(id: String, code: Int, reason: String): String {
        val connection = connections[id] ?: return "Unknown WebSocket connection ID"
        return runCatching {
            if (!checkNotNull(connection.socket).close(code, reason)) {
                "WebSocket close request was rejected"
            } else {
                ""
            }
        }.getOrElse { error -> error.message ?: "Unable to close WebSocket" }
    }

    fun destroy() {
        if (destroyed) return
        destroyed = true
        lynxView = null
        connections.values.forEach { connection ->
            connection.terminal.set(true)
            connection.socket?.cancel()
        }
        connections.clear()
    }

    private fun send(id: String, operation: (WebSocket) -> Boolean): String {
        val connection = connections[id] ?: return "Unknown WebSocket connection ID"
        val socket = connection.socket ?: return "WebSocket is not ready"
        return runCatching {
            if (operation(socket)) "" else "WebSocket send queue rejected the message"
        }.getOrElse { error -> error.message ?: "Unable to send WebSocket message" }
    }

    private fun finish(
        id: String,
        connection: Connection,
        code: Int,
        reason: String,
        wasClean: Boolean,
    ) {
        if (!connection.terminal.compareAndSet(false, true)) return
        connections.remove(id, connection)
        emitClose(id, code, reason, wasClean)
    }

    private fun emitClose(id: String, code: Int, reason: String, wasClean: Boolean) {
        emit(
            id,
            JavaOnlyMap().apply {
                putString("type", "close")
                putInt("code", code)
                putString("reason", reason)
                putBoolean("wasClean", wasClean)
            },
        )
    }

    private fun emit(id: String, payload: JavaOnlyMap) {
        if (destroyed) return
        payload.putString("id", id)
        activity.runOnUiThread {
            if (!destroyed) {
                lynxView?.sendGlobalEvent(EVENT_NAME, JavaOnlyArray.of(payload))
            }
        }
    }

    private fun isAllowedUrl(url: String): Boolean =
        url.startsWith("wss://", ignoreCase = true) ||
            (BuildConfig.DEBUG && url.startsWith("ws://", ignoreCase = true))

    companion object {
        const val EVENT_NAME = "nativeWebSocket"
        private val CONNECTION_ID = Regex("^[A-Za-z0-9._-]{1,128}$")
    }
}

/** Lynx NativeModule facade for the app-owned Android WebSocket controller. */
class NativeWebSocketModule(context: Context, param: Any?) : LynxModule(context, param) {
    private val controller = param as? NativeWebSocketController

    @LynxMethod
    fun connect(options: ReadableMap, callback: Callback) {
        val host = controller
        if (host == null) {
            callback.invoke("Native WebSocket has no Activity host")
            return
        }
        val protocols = options.getArray("protocols")
            ?.asArrayList()
            ?.filterIsInstance<String>()
            .orEmpty()
        val headers = options.getMap("headers")
            ?.asHashMap()
            .orEmpty()
            .mapValues { (_, value) -> value as? String ?: value.toString() }
        callback.invoke(
            host.connect(
                id = options.getString("id", ""),
                url = options.getString("url", ""),
                protocols = protocols,
                headers = headers,
            ),
        )
    }

    @LynxMethod
    fun sendText(id: String, data: String, callback: Callback) {
        callback.invoke(controller?.sendText(id, data) ?: "Native WebSocket has no Activity host")
    }

    @LynxMethod
    fun sendBase64(id: String, data: String, callback: Callback) {
        callback.invoke(controller?.sendBase64(id, data) ?: "Native WebSocket has no Activity host")
    }

    @LynxMethod
    fun close(id: String, code: Int, reason: String, callback: Callback) {
        callback.invoke(controller?.close(id, code, reason) ?: "Native WebSocket has no Activity host")
    }

    companion object {
        const val NAME = "NativeWebSocketModule"
    }
}
